hazelcast/statistics.py

```diff
@@ -0,0 +1,251 @@
+        self._statistics_timer = None
+        self._failed_gauges = set()
+
+    def start(self):
+        if not self._enabled:
+            return
+
+        period = self._client.properties.get_seconds(ClientProperties.STATISTICS_PERIOD_SECONDS)
+        if period <= 0:
+            default_period = self._client.properties.get_seconds_positive_or_default(
+                ClientProperties.STATISTICS_PERIOD_SECONDS)
+
+            self.logger.warning("Provided client statistics {} cannot be less than or equal to 0."
+                                "You provided {} as the configuration. Client will use the default value "
+                                "{} instead.".format(ClientProperties.STATISTICS_PERIOD_SECONDS.name,
+                                                     period,
+                                                     default_period))
+            period = default_period
+
+        def _statistics_task():
+            self._send_statistics()
+            self._statistics_timer = self._client.reactor.add_timer(period, _statistics_task)
+
+        self._statistics_timer = self._client.reactor.add_timer(period, _statistics_task)
+
+        self.logger.info("Client statistics enabled with the period of {} seconds.".format(period))
+
+    def shutdown(self):
+        if self._statistics_timer:
+            self._statistics_timer.cancel()
+
+    def _send_statistics(self):
+        owner_connection = self._get_owner_connection()
+        if owner_connection is None:
+            self.logger.debug("Cannot send client statistics to the server. No owner connection.")
+            return
+
+        stats = []
+        self._fill_metrics(stats, owner_connection)
+        self._add_near_cache_stats(stats)
+        self._add_runtime_and_os_stats(stats)
+        self._send_stats_to_owner("".join(stats), owner_connection)
+
+    def _send_stats_to_owner(self, stats, owner_connection):
+        request = client_statistics_codec.encode_request(stats)
+        self._client.invoker.invoke_on_connection(request, owner_connection)
+
+    def _add_runtime_and_os_stats(self, stats):
+        os_and_runtime_stats = self._get_os_and_runtime_stats()
+        for stat in os_and_runtime_stats:
+            self._add_stat(stats, stat, os_and_runtime_stats[stat])
+
+    def _get_os_and_runtime_stats(self):
+        psutil_stats = {}
+        if PSUTIL_ENABLED:
+            try:
+                # Availability: All platforms. Try to be sure that it is working.
+                if "phy_mem" not in self._failed_gauges:
+                    memory_info = psutil.virtual_memory()
+                    psutil_stats["os.totalPhysicalMemorySize"] = memory_info.total
+                    psutil_stats["os.freePhysicalMemorySize"] = memory_info.available
+            except Exception as ex:
+                self.logger.warning(self._EXCEPTION_STR.format("physical memory size", ex))
+                self._failed_gauges.add("phy_mem")
```

Comment:
See https://github.com/giampaolo/psutil/blob/master/psutil/_pslinux.py#L478 for example. When available memory cannot be determined, psutil will not return it so we can get AttributeError here.

Reply:
See https://github.com/giampaolo/psutil/blob/master/psutil/_pslinux.py#L402 here for example. When the `/proc/meminfo` doesn't contain the fields, we can get a KeyError. I think it is best to call them within the try catch block to avoid such edge cases. try catch block adds a very little overhead.